### DISCUSSION TITLE
Testing for child nodes before calling ITerm.LoadAsync()?

### DISCUSSION THREAD

**Discussion by aaronmorgan** [pnp/pnpcore#917] [Q&A] [answered]
Hi, I'm migrating old PnP.Framework code to .Net 6.0 and using the PnP Core SDK Nuget package (1.7.0).

In the earlier incarnation I see the code performed a check of `Term.TermsCount` before using the context to the child terms.
What I can see in the code for PnP Core and GitHub source is we need to load the terms before being able to determine if there are any, e.g.

```
await term.LoadAsync(p => p.Terms);
```

This requires a round trip to the SPO site and in my testing this call takes ~225ms to complete. The parent group contains 3000+ terms and over 11 minutes to check the terms for the possibility of child nodes.

Is there a faster way to achieve this? Have I missed something?

Thank you for your answer.

**Comment by jansenbe** [comment, marked as answer]
@aaronmorgan : Using batching and loading the child terms upfront you can build a fast enumeration. Below code snippet managed to load a very deep nested and large termset in a matter of 10 seconds.

```csharp
var termset = context.TermStore.GetTermSetById("4b000117-03c4-4b2b-81f4-21e2ab26d6be", p => p.Description, p => p.Terms);

// recursively load the terms in the termset
await LoadTermsAsync(termset.Terms);

private async Task LoadTermsAsync(ITermCollection terms)
{
    var batch = terms.PnPContext.NewBatch();

    foreach (var term in terms.AsRequested())
    {
        await term.LoadBatchAsync(batch, p => p.Labels, p => p.Terms);
    }

    await terms.PnPContext.ExecuteAsync(batch);

    foreach (var term in terms.AsRequested())
    {
        if (term.Terms.AsRequested().Count() > 0)
        {
            // Load the possible child terms
            await LoadTermsAsync(term.Terms);
        }
    }
}
```

**Reply by aaronmorgan** [reply]
@jansenbe thank you, yes that has driven the execution time down a lot, it's now about 60% faster.

I wonder if it's now hitting throughput limits (based on the documented 'Batch Limits') as there are 3050 Terms at the root of this TermGroup and while none of them have children that'll cause the batch to be split numerous times before going to SharePoint.

A.

**Reply by jansenbe** [reply]
A graph batch can only contain 20 requests, we'll automatically split them. If you've a very large set of level 1 nodes then the optimization is not super, but at least you're able to have 1 instead of 20 requests @aaronmorgan
